[TEST] RecoveryIT.testHistoryUUIDIsGenerated occasionally fails on 6.x

Build: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-windows-compatibility/1119/console

Reproduce with: 

```
gradlew :qa:rolling-upgrade:v5.6.10-SNAPSHOT#upgradedClusterTestRunner -Dtests.seed=CC9DDAFEF7C7362F -Dtests.class=org.elasticsearch.upgrades.RecoveryIT -Dtests.method="testHistoryUUIDIsGenerated" -Dtests.security.manager=true -Dtests.locale=en-IN -Dtests.timezone=Australia/Eucla
```

Failure:

```
RecoveryIT.testHistoryUUIDIsGenerated <<< FAILURES!
07:49:28    > Throwable #1: java.lang.AssertionError: different history uuid found for shard on BPdfH1zoREy4bR0ga9iksg
07:49:28    > Expected: "D9AAXTrJT7WOJsynoFB0LQ"
07:49:28    >      but: was "JsLgmYhvTEyjm_x9NWESGQ"
07:49:28    > 	at __randomizedtesting.SeedInfo.seed([CC9DDAFEF7C7362F:E43FE5891A4AA329]:0)
07:49:28    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
07:49:28    > 	at org.elasticsearch.upgrades.RecoveryIT.testHistoryUUIDIsGenerated(RecoveryIT.java:79)
07:49:28    > 	at java.lang.Thread.run(Thread.java:748)
```

I could not reproduce locally.

--- a/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java
+++ b/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java
@@ -60,7 +60,11 @@ public class RecoveryIT extends AbstractRollingTestCase {
                 // then delayed allocation will kick in. When the node comes back, the master will search for a copy
                 // but the recovering copy will be seen as invalid and the cluster health won't return to GREEN
                 // before timing out
-                .put(INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), "100ms");
+                .put(INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), "100ms")
+                // if this index can be allocated to any testing nodes, the rebalance allocation decider may relocate
+                // this index. if the relocation from 5.x to 6.x, we will have two different history_uuids because
+                // history_uuid is bootstrapped twice by two different 6.x nodes: a replica and a relocating primary.
+                .put("index.routing.allocation.include._name", "*node-0,*node-1");
             createIndex(index, settings.build());
         } else if (CLUSTER_TYPE == ClusterType.UPGRADED) {
             ensureGreen(index);

--- a/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java
+++ b/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java
@@ -49,7 +49,6 @@ import static org.hamcrest.Matchers.notNullValue;
  */
 public class RecoveryIT extends AbstractRollingTestCase {
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/31291")
     public void testHistoryUUIDIsGenerated() throws Exception {
         final String index = "index_history_uuid";
         if (CLUSTER_TYPE == ClusterType.OLD) {

